pkg/urbit/worker/mars.c

```diff
@@ -696,272 +700,273 @@ _mars_timer_cb(uv_timer_t* tim_u)
 
-  u3l_log("--------------- bootstrap complete ----------------\r\n");
-  return c3y;
+  u3e_save();
+
+  goto succeed;
+
+fail:
+  return 0;
+
+succeed:
+  return 1;
 }
 
-/* u3_mars_init(): init mars, replay if necessary.
-*/
+//! @n (1) XX load/set secrets.
+//! @n (2) Play boot events if necessary.
+//! @n (3) Turn on async commit mode for last epoch.
+//! @n (4) These must be set after the replay in order to adhere to the
+//!        invariant that all noun references be roots when calling the u3_saga
+//!        interface.
+//! @n (5) Send ready status message. XX version negotation.
+//! @n (6) XX check return, make interval configurable.
 u3_mars*
-u3_mars_init(c3_c*    dir_c,
-             u3_moat* inn_u,
-             u3_mojo* out_u,
-             c3_d     eve_d)
+u3_mars_init(c3_c* dir_c, u3_moat* inn_u, u3_mojo* out_u, c3_d eve_d)
 {
-  u3_mars* mar_u = c3_malloc(sizeof(*mar_u));
-  mar_u->dir_c = dir_c;
-  mar_u->inn_u = inn_u;
-  mar_u->out_u = out_u;
-  mar_u->sen_d = mar_u->dun_d = u3A->eve_d;
-  mar_u->mug_l = u3r_mug(u3A->roc);
-  mar_u->pac_o = mar_u->rec_o = mar_u->mut_o = c3n;
-  mar_u->sac   = u3_nul;
-  mar_u->sat_e = u3_mars_work_e;
-  mar_u->gif_u.ent_u = mar_u->gif_u.ext_u = 0;
-  mar_u->xit_f = 0;
-
-  mar_u->sil_u = u3s_cue_xeno_init();
-
-  //  initialize persistence
-  //
-  //    XX load/set secrets
-  //
-  if ( !(mar_u->log_u = u3_disk_init(dir_c)) ) {
-    fprintf(stderr, "mars: disk init fail\r\n");
-    c3_free(mar_u);
-    return 0;
+  u3_mars* mar_u;
+  {
+    mar_u  = c3_calloc(sizeof(*mar_u));
+    *mar_u = (u3_mars){
+      .dir_u = c3_path_fv(1, dir_c),
+      .sen_d = u3A->eve_d,
+      .dun_d = u3A->eve_d,
+      .pac_o = c3n,
+      .rec_o = c3n,
+      .mut_o = c3n,
+      .sac   = u3_nul,
+      .inn_u = inn_u,
+      .out_u = out_u,
+      .sat_e = u3_mars_work_e,
+      .gif_u = c3_list_init(),
+    };
   }
 
-  if ( c3n == u3_disk_read_meta(mar_u->log_u, &(mar_u->met_u)) ) {
-    fprintf(stderr, "mars: disk meta fail\r\n");
-    u3_disk_exit(mar_u->log_u);
-    c3_free(mar_u);
-    return 0;
-  }
+  c3_path* const lok_u = c3_path_fp(mar_u->dir_u);
+  u3_lock_acquire(lok_u);
 
-  if ( !mar_u->dun_d ) {
-    if ( c3n == _mars_do_boot(mar_u->log_u, mar_u->met_u.lif_w) ) {
-      fprintf(stderr, "mars: boot fail\r\n");
-      u3_disk_exit(mar_u->log_u);
-      c3_free(mar_u);
-      return 0;
+  { // (1)
+
+    if ( eve_d && eve_d <= mar_u->dun_d ) {
+      fprintf(stderr,
+              "mars: replay to %" PRIu64 " already done (at %" PRIu64 ")\r\n",
+              eve_d,
+              mar_u->dun_d);
+      goto release_lock;
     }
 
-    mar_u->sen_d = mar_u->dun_d = mar_u->met_u.lif_w;
-    u3e_save();
-  }
+    c3_path_push(mar_u->dir_u, ".urb");
+    c3_path_push(mar_u->dir_u, "log");
+    try_saga(mar_u->log_u = u3_saga_open(mar_u->dir_u, &mar_u->met_u),
+             goto release_lock,
+             "failed to open event log at %s",
+             c3_path_str(mar_u->dir_u));
```

Comment:
Perhaps unsurprisingly, I disagree. Each module that has a `try_<module_name>` macro (i.e. `saga`, `epoc`, etc) uses consistent return types in its interface. That is, with few exceptions, zero conveys failure and non-zero conveys success. As a result, the majority of use cases of these interfaces obeys a well-defined pattern. Furthermore, each of the `try_*` macros work in exactly the same way. So the use of the macros reduces the cognitive energy required to read the code because there are fewer moving pieces around a call into one of these interfaces. Your example is harder to read because there's more slack i.e. more spaces in which arbitrary code can be inserted, which makes it harder to understand. Using `try_saga`, we have
```c
try_saga(mar_u->log_u = u3_saga_open(mar_u->dir_u, &mar_u->met_u), goto release_lock, /* snip */);
```
which unambiguously reads as "attempt to open the event log and assign it to `mar_u->log_u`, and if that fails, `goto release_lock`. In other words, the `try_*` macros reduce the number of branches that someone reading the code needs to hold in their head at any one time because many of the branches are identical, which allows them to be collapsed into a single pattern by the relevant `try_*` macro. And in the case where more complex logic is required, an `if` statement as you describe will always work *and*, by breaking the `try_*` pattern, will obviously convey that something different is going on.